### ISSUE TITLE
Wine Taskbar Icons Incorrect

### ISSUE THREAD

**Issue by greyskier** [linuxmint/Cinnamon#3814]
Applications running in wine show the incorrect icon in the taskbar. Instead of showing the running application in wine they just show the wine Windows icon. They show the correct icon in the menu because the correct icon is listed in the  .desktop file at ~/.local/share/applications.
I'm using LinuxMint 17.1 Cinnamon and this wasn't an issue when I was using the MATE desktop. This issue was previously closed but it is not fixed as several other people left comments after it was closed but apparently didn't create a new thread. This is a real PITA for those of us who have to run several Windows apps at once along with Firefox, file manager and Gedit etc. I have spent hours reading and trying different things to no avail.

**Comment by manjukirans** [comment]
This issue seems to be there since a long time. Can we reopen the earlier bug?
https://github.com/linuxmint/Cinnamon/issues/581
mtwebster had closed this bug due to inactivity.

**Comment by igordovgaluk** [comment]
@greyskier @manjukirans i have same issue with Far Manager. On the other hand WinMerge is showed with correct icon. so it could be Wine's bug. 
What apps have wrong icons in your case?

**Comment by greyskier** [comment]
I do genealogy and I need to have several windows open at once for Firefox,file manager, gedit, The Master Genealogist(Windows program), Dynamic Famiy Tree Compiler(Windows program), Gedpage(Windows program), plus usually Rhythmbox music player and Thunderbird.  This makes the task bar entries very small and the Windows programs all end up resembling Firefox as their icons don't display properly making switching between them more difficult as you have to stop and look carefully.

**Comment by manjukirans** [comment]
Source Insight, Microsoft Office Applications, Teamviewer are the applications I have used and seen the problem. The reason this could be a Cinnamon issue rather than wine issue is because, this issue does not show up in MATE and Gnome. Also, if you use a dock (such as docky), the issue is not seen,

**Comment by Naruki** [comment]
Also test any apps that run in a terminal, such as HTOP or Midnight Commander. Those are the two that cause this problem for me.

**Comment by igordovgaluk** [comment]
issue with Far Manager resolved by using winconsole instead of wine to launch the app

**Comment by JosephMcc** [comment]
Could you test this again once you have Cinnamon 2.8.3 or greater.

**Comment by igordovgaluk** [comment]
checked with 2.8.3-20151111040009-trusty, HTOP, Midnight Commander, TeamViewer 10, Far Manager (wine) - not fixed

**Comment by JosephMcc** [comment]
>Also test any apps that run in a terminal, such as HTOP or Midnight Commander. Those are the two that cause this problem for me.

Don't they show up in the window list as a terminal? Do they behave differently in other DE's somehow?

**Comment by greyskier** [comment]
It appears to be using the icon /usr/share/icons/Mint-X/apps/xx/wine.png rather than the one that is in the .desktop file

**Comment by ManIVIctorious** [comment]
I can confirm this is still an issue in cinnamon 3.0.7

**Comment by gstavi** [comment]
Copying my comment from [#581](https://github.com/linuxmint/Cinnamon/issues/581)

Maybe that will help:

I had similar issue with Notepad++ under wine under ubuntu with gnome-desktop.
Correct Notepad++ icon would show in 'Applications' menu/launcher but Alt-Tab showed default wine icon.

The solution I found:
My Notepad++ executable is called npp.exe.
Under ~/.local/share/applications I had a shortcut called npp.desktop with reference to correct icon.

I RENAMED npp.desktop TO npp.exe.desktop AND PROBLEM SOLVED.

It seems that Alt-Tab takes the "process name" and uses it to lookup a desktop file for the icon. Make sure the desktop file name matches the process name.

Perhaps this lookup procedure should have the ability to strip common windows extensions (namely .exe) from the name.

**Comment by kvnxiao** [comment]
Is this ever going to be fixed?

**Comment by GitHubIsToxic** [comment]
Still not fixed in Mint 19.

**Comment by icarter09** [comment]
@GitHubIsToxic is this still an issue with Mint 19.3?

**Comment by claudiux** [comment]
@icarter09 Yes, this is still an issue with Mint 19.3. Just a black icon appears in the taskbar, instead of the app icon.

**Comment by GitHubIsToxic** [comment]
Running Mint 19 Cinnamon 3.8.9 here; yes all Wine applications have the Wine icon (red square with Windows XP-ish logo).

![taskbar_icons](https://user-images.githubusercontent.com/6215449/71262831-e7ad6100-2340-11ea-9dab-bcac3e90f6fa.png)

For me it is not black, like @claudiux describes, although I vaguely recall see something like it years ago.

IIRC in previous Mint versions the icon in taskbar was the Wine glass icon instead of the current windows icon.

**Comment by icarter09** [comment]
@GitHubIsToxic @claudiux: I haven't used Wine in awhile, but I thought the Windows icon in the image provided by @GitHubIsToxic was the correct icon. (Stupid question) Has this been tested lately in any other OS or DE?

**Comment by GitHubIsToxic** [comment]
I seem to recall icons were displayed properly on Ubuntu (Gnome 2 or XFCE), after Unity happened (and thus the taskbar disappeared) I switched to Mint. Somehow I know it's wrong in Mint, anyway.

**Comment by anhollander516** [comment]
> Copying my comment from [#581](https://github.com/linuxmint/Cinnamon/issues/581)
> 
> Maybe that will help:
> 
> I had similar issue with Notepad++ under wine under ubuntu with gnome-desktop. Correct Notepad++ icon would show in 'Applications' menu/launcher but Alt-Tab showed default wine icon.
> 
> The solution I found: My Notepad++ executable is called npp.exe. Under ~/.local/share/applications I had a shortcut called npp.desktop with reference to correct icon.
> 
> I RENAMED npp.desktop TO npp.exe.desktop AND PROBLEM SOLVED.
> 
> It seems that Alt-Tab takes the "process name" and uses it to lookup a desktop file for the icon. Make sure the desktop file name matches the process name.
> 
> Perhaps this lookup procedure should have the ability to strip common windows extensions (namely .exe) from the name.

This is still broken on Linux Mint 22.2, many years later.

This was part of the solution. The second part of the solution was to set `StartupWMClass` in the desktop file to the window class as returned by `xprop WM_CLASS` (which normally is the executable name, but there may be cases where it isn't? Not sure.).

**Comment by leigh123linux** [comment]
> Copying my comment from [#581](https://github.com/linuxmint/Cinnamon/issues/581)
> 
> Maybe that will help:
> 
> I had similar issue with Notepad++ under wine under ubuntu with gnome-desktop. Correct Notepad++ icon would show in 'Applications' menu/launcher but Alt-Tab showed default wine icon.
> 
> The solution I found: My Notepad++ executable is called npp.exe. Under ~/.local/share/applications I had a shortcut called npp.desktop with reference to correct icon.
> 
> I RENAMED npp.desktop TO npp.exe.desktop AND PROBLEM SOLVED.
> 
> It seems that Alt-Tab takes the "process name" and uses it to lookup a desktop file for the icon. Make sure the desktop file name matches the process name.
> 
> Perhaps this lookup procedure should have the ability to strip common windows extensions (namely .exe) from the name.

Shouldn't wine be responsible for generating valid desktop files that comply to the freedesktop standards?
